Add unified /api/healthz

spencer-p opened the pull request:
## Why are these changes needed?

This new endpoint on the HealthzAgent class combines the status of /api/local_raylet_healthz and /api/gcs_healthz into one endpoint for use with Kubernetes.

<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->



<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

 See #56204.
<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

spencer-p committed:
Add unified /api/healthz

This new endpoint on the HealthzAgent class combines the status of
/api/local_raylet_healthz and /api/gcs_healthz into one endpoint for use
with Kubernetes. See #56204.

Signed-off-by: Spencer Peterson <spencerjp@google.com>

spencer-p committed:
fix 'is not' -> != for int

Signed-off-by: Spencer Peterson <spencerjp@google.com>

edoakes commented on python/ray/dashboard/modules/reporter/healthz_agent.py:
```diff
@@ -52,6 +52,26 @@ async def health_check(self, req: Request) -> Response:
         )
 
+    @routes.get("/api/healthz")
+    async def unified_health(self, req: Request) -> Response:
```
Suggest to:

- Always check the health of both (concurrently). That way we know if only raylet is unhealthy or GCS is too. As-is, you wouldn't know if GCS is healthy.
- Warning log if health check fails for either of them.

edoakes commented on python/ray/dashboard/modules/reporter/healthz_agent.py:
```diff
@@ -52,6 +52,26 @@ async def health_check(self, req: Request) -> Response:
+        if raylet_resp.status != 200:
+            return raylet_resp
+
+        # Check GCS health.
+        try:
+            gcs_alive = await self._health_checker.check_gcs_liveness()
+            if not gcs_alive:
+                return Response(status=503, text="GCS health check failed.")
+        except Exception as e:
+            return Response(status=503, text=f"GCS health check failed: {e}")
```
dedup'ing into a method/attribute on the `DashboardAgentModule` class would be great 👍

spencer-p committed:
Add DashboardAgent is_head property

Signed-off-by: Spencer Peterson <spencerjp@google.com>

spencer-p committed:
Check health concurrently, no remote GCS check

Signed-off-by: Spencer Peterson <spencerjp@google.com>

spencer-p committed:
Refactor unified health check; add logging

Signed-off-by: Spencer Peterson <spencerjp@google.com>

spencer-p committed:
py format

Signed-off-by: Spencer Peterson <spencerjp@google.com>

spencer-p committed:
Fix unified health check

- imports
- Use asyncio.gather instead of TaskGroup, which is not available on
  3.10
- Fix missing happy path in GCS check

Signed-off-by: Spencer Peterson <spencerjp@google.com>

edoakes commented on python/ray/dashboard/agent.py:
```diff
@@ -163,6 +163,10 @@ def http_session(self):
     def get_node_id(self) -> str:
         return self.node_id
 
+    @property
+    def is_head(self) -> bool:
+        return self.ip == parse_address(self.gcs_address)[0]
```
I think this is not reliable if people are using the ray `cluster_utils` utility to run multiple logical nodes on the same host

The most reliable identity check would be using the node ID

edoakes commented on python/ray/dashboard/modules/reporter/healthz_agent.py:
```diff
@@ -52,6 +54,41 @@ async def health_check(self, req: Request) -> Response:
             content_type="application/text",
         )
 
+    async def local_gcs_health(self) -> Response:
+        # Check GCS health, unless there is no local GCS.
+        if not self._dashboard_agent.is_head:
+            return Response(status=200, text="success (no local gcs)")
+        try:
+            gcs_alive = await self._health_checker.check_gcs_liveness()
+            if not gcs_alive:
+                return Response(status=503, text="GCS health check failed.")
+        except Exception as e:
+            return Response(status=503, text=f"GCS health check failed: {e}")
+        return Response(status=200, text="success")
```
I would just raise an exception from this & the raylet health check rather than do the awkward status code aggregation logic

might need to refactor raylet health check to call into an internal method that you share here.

edoakes reviewed (commented):
Let's add some basic tests. If it's possible to write integration tests for this, that would be great, but not sure how because if the raylet and/or GCS become unhealthy, I think the agent will shortly crash...

Else we can scaffold up some unit tests (without too much mocking please)

edoakes commented:
unstale

spencer-p committed:
Dashboard agent fetches is_head from GCS

Signed-off-by: Spencer Peterson <spencerjp@google.com>

spencer-p committed:
Track health check failures w/ exceptions

Signed-off-by: Spencer Peterson <spencerjp@google.com>

spencer-p committed:
lint

Signed-off-by: Spencer Peterson <spencerjp@google.com>

spencer-p committed:
Unified health check tests

fixes missing await and bad logging references.

Signed-off-by: Spencer Peterson <spencerjp@google.com>

edoakes commented on python/ray/dashboard/agent.py:
```diff
@@ -76,6 +76,12 @@ def __init__(
+        node_info = self.gcs_client.get_all_node_info().get(
+            NodeID.from_hex(self.node_id)
+        )
+        self.is_head = node_info is not None and node_info.is_head_node
```
I don't think this is true

edoakes replied:
but allowing the `node_info is None` case to silently pass is sketchy

can you dig a bit to see if this is possible? I think we need to check the startup flow in the raylet. check `raylet/main.cc` for ordering between registering w/ GCS and starting the agent

spencer-p replied:
hm it looks like the dashboard agent is started in NodeManager's constructor which precedes registering with GCS in NodeManager::Start. I'll add an assert so we crash if that happens.

which shouldn't be any worse even if i've incorrectly traced this anyway

edoakes replied:
How about adding a short timeout/wait loop? Give a 30s grace period for the node info to be populated, if it doesn't show up, then crash (should never happen)

spencer-p replied:
Done. lmk if you would prefer to see call_with_retry used instead - it seemed more verbose in this case to me.

spencer-p committed:
dashboard asserts its own nodeinfo exists

Signed-off-by: Spencer Peterson <spencerjp@google.com>

spencer-p committed:
retry node info fetch for 30s at most

Signed-off-by: Spencer Peterson <spencerjp@google.com>

spencer-p commented on python/ray/dashboard/modules/reporter/healthz_agent.py:
```diff
@@ -45,10 +61,37 @@ async def health_check(self, req: Request) -> Response:
+            return "success (no local gcs)"
+        gcs_alive = await self._health_checker.check_gcs_liveness()
+        if not gcs_alive:
+            raise Exception("GCS health check failed.")
```
this is true but I will leave as-is since future editors of check_gcs_liveness may not know the return value is unused

400Ping reviewed (approved):
Nice.

spencer-p committed:
switch for loop for call_with_retry

Signed-off-by: Spencer Peterson <spencerjp@google.com>

spencer-p commented on python/ray/dashboard/agent.py:
```diff
@@ -76,6 +77,18 @@ def __init__(
 
+        # Fetch node info and save is_head.
+        node_info = None
+        for _ in range(30):
```
the for loop is not the most readable tbh and it seems there is an easy way to use call_with_retry if I let dict indexing raise a keyerror. changed
